pkg/storegateway/indexheader/lazy_binary_reader_test.go

```diff
@@ -294,6 +297,63 @@ func testLazyBinaryReader(t *testing.T, bkt objstore.BucketReader, dir string, i
 		return NewStreamBinaryReader(ctx, logger, bkt, dir, id, 3, NewStreamBinaryReaderMetrics(nil), Config{})
 	}
 
-	reader, err := NewLazyBinaryReader(ctx, factory, logger, bkt, dir, id, NewLazyBinaryReaderMetrics(nil), nil)
+	reader, err := NewLazyBinaryReader(ctx, factory, logger, bkt, dir, id, NewLazyBinaryReaderMetrics(nil), nil, gate.NewNoop())
 	test(t, reader, err)
 }
+
+// Tests if LazyBinaryReader blocks concurrent loads such that it doesn't pass the configured maximum
+func TestLazyBinaryReader_ShouldBlockMaxConcurrency(t *testing.T) {
+	ctx := context.Background()
+
+	tmpDir := filepath.Join(t.TempDir(), "test-indexheader")
+	bkt, err := filesystem.NewBucket(filepath.Join(tmpDir, "bkt"))
+	require.NoError(t, err)
+	t.Cleanup(func() { require.NoError(t, bkt.Close()) })
+
+	// Create block.
+	blockID, err := block.CreateBlock(ctx, tmpDir, []labels.Labels{
+		labels.FromStrings("a", "1"),
+		labels.FromStrings("a", "2"),
+		labels.FromStrings("a", "3"),
+	}, 100, 0, 1000, labels.FromStrings("ext1", "1"))
+	require.NoError(t, err)
+	require.NoError(t, block.Upload(ctx, log.NewNopLogger(), bkt, filepath.Join(tmpDir, blockID.String()), nil))
+
+	logger := log.NewNopLogger()
+	var total int
```

Comment:
Since these are being updated from multiple goroutines at the same time, they need to be [atomics](https://pkg.go.dev/go.uber.org/atomic?utm_source=godoc):

```
total := atomic.NewUint32(0)
```